DOCSP-37536 Bulk Write Operations (#48)
mongodb/docs-pymongo: source/write/bulk-write.txt

.. _pymongo-bulk-write:

=====================
Bulk Write Operations
=====================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: insert, update, replace, code example

Overview
--------

> In this guide, you can learn how to use bulk write operations with
> {+driver-short+}. You can perform multiple write operations
> in a single database call by using bulk operations.
> 
> Consider a scenario in which you want to insert a document into a collection,
> update multiple other documents, then delete a document. Performing
> these operations by using their individual methods makes a call to the database
> for each method. You can instead use bulk write operations to perform
> these operations in a single database call.

Reviewer: ```suggestion
Consider a scenario in which you want to insert a document into a collection,
update multiple other documents, then delete a document. If you use
individual methods, each operation requires its own database call. This guide
shows you how to use bulk write operations to perform multiple write operations
in a single database call. 
```

> Performing Bulk Operations

Reviewer: This title seems a little vague and is basically the point of the entire page. What if we called it something like "Define the Write Operations"?

--------------------------

> To perform a bulk write operation, you must first create an instance of one of
> the following operation classes for each write operation you want to perform:

Reviewer: ```suggestion
For each write operation you want to perform, create an instance of one of
the following operation classes:
```

- ``InsertOne``
- ``UpdateOne``
- ``UpdateMany``
- ``ReplaceOne``
- ``DeleteOne``
- ``DeleteMany``

Then, pass a list of these instances to the ``bulk_write()`` method.

> The following sections show how to create and use operation classes for
> different write operations.

Reviewer: S:

```suggestion
The following sections show how to create and use instances of the
preceding classes.
```

Insert Operations
~~~~~~~~~~~~~~~~~

To perform an insert operation, create an instance of ``InsertOne`` and specify

> the document you want to insert. To insert multiple documents, you must create

Reviewer: ```suggestion
the document you want to insert. To insert multiple documents, create
```

> the document you want to insert. To insert multiple documents, you must create

Reviewer: S: Consider putting this bit about multiple inserts after the code example

an instance of ``InsertOne`` for each document.

The following example creates an instance of ``InsertOne``:

> .. literalinclude:: /includes/write/bulk-write.py

Author: All code examples are currently placeholders

   :start-after: start-bulk-insert-one
   :end-before: end-bulk-insert-one
   :language: python
   :copyable:

Update Operations
~~~~~~~~~~~~~~~~~

> To perform a bulk update operation, create an instance of ``UpdateOne`` or
> ``UpdateMany`` and pass it a **query filter** and the update operation you want
> to perform. A query filter specifies the criteria used to match documents to
> in your collection.
> 
> ``UpdateOne`` updates *the first* document that matches your query filter.
> ``UpdateMany`` updates *all* documents that match your query filter.
> 
> The following example creates an instance of ``UpdateOne``:
> 
> .. literalinclude:: /includes/write/bulk-write.py
>    :start-after: start-bulk-update-one
>    :end-before: end-bulk-update-one
>    :language: python
>    :copyable:

Reviewer: S: Dividing this into two sections probably makes it easier to find what you're looking for at a glance. Maybe something like this (but feel free to edit)

```suggestion
To update a document, create an instance of ``UpdateOne`` and pass in
the following arguments:

- A query filter that specifies the document to update. ``UpdateOne`` updates *the first document* that matches this query filter.
- The update operation you want to perform [in what format].

The following example creates an instance of ``UpdateOne``:

.. literalinclude:: /includes/write/bulk-write.py
   :start-after: start-bulk-update-one
   :end-before: end-bulk-update-one
   :language: python
   :copyable:
   
   To update multiple documents, create an instance of ``UpdateMany``
   and pass in the same arguments. ``UpdateMany`` updates *all* documents
    that match your query filter.

```

The following example creates an instance of ``UpdateMany``:

.. literalinclude:: /includes/write/bulk-write.py
   :start-after: start-bulk-update-many
   :end-before: end-bulk-update-many
   :language: python
   :copyable:

Replace Operations
~~~~~~~~~~~~~~~~~~

> Replace operations remove all fields and values of a specified document and
> replaces them with new ones. To perform a replace operation, create an instance

Reviewer: ```suggestion
A replace operation removes all fields and values of a specified document and
replaces them with new ones. To perform a replace operation, create an instance
```

of ``ReplaceOne`` and pass it a query filter and the fields and values you want

> to replace with. To replace multiple documents, you must create

Reviewer: ```suggestion
to store in the matching document. To replace multiple documents, you must create
```

> an instance of ``ReplaceOne`` for each document.

Reviewer: S: Consider putting after the code example

The following example creates an instance of ``ReplaceOne``:

.. literalinclude:: /includes/write/bulk-write.py
   :start-after: start-bulk-replace-one
   :end-before: end-bulk-replace-one
   :language: python
   :copyable:

Delete Operations
~~~~~~~~~~~~~~~~~

> To perform a delete operation, create an instance of ``DeleteOne`` or ``DeleteMany`` and pass it a

Reviewer: S: Would probably follow a similar format here as for UpdateOne and UpdateMany above

query filter specifying the documents you want to delete.

``DeleteOne`` removes *the first* document that matches your query filter.
``DeleteMany`` removes *all* documents that match your query filter.

The following example creates an instance of ``DeleteOne``:

.. literalinclude:: /includes/write/bulk-write.py
   :start-after: start-bulk-delete-one
   :end-before: end-bulk-delete-one
   :language: python
   :copyable:

The following example creates an instance of ``DeleteMany``:

.. literalinclude:: /includes/write/bulk-write.py
   :start-after: start-bulk-delete-many
   :end-before: end-bulk-delete-many
   :language: python
   :copyable:

> Mixed Bulk Write Operations

Reviewer: Maybe:

```suggestion
Call the bulk_write() Method
```

I'm thinking if the section titles are basically the steps to follow, the user can look at On This Page and see an overview of what to do

> ----------------------------

Author: This section is the "main example" of how to perform a bulk write. I put it farther down the page because you need to use the operations described in the earlier sections to be able to use it. However, this puts the example 3/4 down the page, which users have complained about in the past. Should we move this section up and note that it's explained later? Or maybe add another example in the overview? Open to suggestions here.

Reviewer: Since we're separating usage examples by topic and putting them right up front, I think it's okay if longer code examples are farther down the page in "Fundamentals."

> You can perform multiple types of write operations in a single ``bulk_write()``
> method call. To do so, pass a list of operation class instances to the
> ``bulk_write()`` method. By default, the method runs each operation in the order
> it's provided.

Reviewer: ```suggestion
After you define a class instance for each operation you want to perform,
pass a list of these instances to the ``bulk_write()`` method.
By default, the method runs the operations in the order
they're defined in the list.
```

The following example performs multiple write operations by using the
``bulk_write()`` method:

.. literalinclude:: /includes/write/bulk-write.py
   :start-after: start-bulk-write-mixed
   :end-before: end-bulk-write-mixed
   :language: python
   :copyable:

If any of the write operations fail, {+driver-short+} raises a
``BulkWriteError`` and does not perform any further operations.
``BulkWriteError`` provides a ``details`` attribute that includes the operation

> that failed, and details about the exception.

Reviewer: Move to Troubleshooting section at page end?

Author: I put this here since it only applies to an ordered bulk write shown in this section example, not to ``bulk_write`` in general. It seems a little more informational than troubleshooting imo but I can move it if you feel strongly

.. note::

>    When PyMongo runs a bulk operation, it uses the ``write_concern`` of the

Reviewer: ```suggestion
   When {+driver-short+} runs a bulk operation, it uses the ``write_concern`` of the
```

   collection in which the operation is running. The driver reports all write
   concern errors after attempting all operations, regardless of execution order. 

Customize Bulk Write Operations
-------------------------------

The ``bulk_write()`` method optionally accepts additional
parameters, which represent options you can use to configure the bulk write
operation. If you don't specify any additional options, the driver does not customize
the bulk write operation.

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Property
     - Description

   * - ``ordered``
     - | If ``True``, the driver performs the write operations in the order
         provided. If an error occurs, the remaining operations are not

>          attempted. If ``False``, the driver performs the operations in an
>          arbitrary order and attempts to perform each operation.
>        | Defaults to ``True``.

Reviewer: ```suggestion
         attempted. 
       | 
       | If ``False``, the driver performs *all* operations, in an
         arbitrary order, even if an error occurs.
       | Defaults to ``True``.
```

   * - ``bypass_document_validation``
     - | Specifies whether the operation bypasses document-level validation. For more
         information, see :manual:`Schema
         Validation </core/schema-validation/#schema-validation>` in the MongoDB
         Server manual.
       | Defaults to ``False``.

   * - ``session``
     - | An instance of ``ClientSession``. For more information, see the `API

>          documentation <{+api-root+}pymongo/client_session.html#pymongo.client_session.ClientSession>`__.

Reviewer: Could put this link under API Documentation at page end and leave it out here

Author: I decided to put this here instead on this page so that the API documentation section would have links to just things that are directly `bulk_write()` related. I guess this technically is related to it, but its just a general option so not really bulk write specific

   * - ``comment``
     - | A comment to attach to the operation. For more information, see the :manual:`delete command
         fields </reference/command/delete/#command-fields>` guide in the

>          MongoDB Server manual for more information.

Reviewer: ```suggestion
         MongoDB Server manual.
```

   * - ``let``
     - | A map of parameter names and values. Values must be constant or closed
         expressions that don't reference document fields. For more information,
         see the :manual:`let statement
         </reference/command/delete/#std-label-delete-let-syntax>` in the
         MongoDB Server manual.

> The following example uses the ``bulk_write()`` with the ``ordered`` options set

Reviewer: ```suggestion
The following example uses the ``bulk_write()`` method with the ``ordered`` options set
```

to ``False``:

.. literalinclude:: /includes/write/bulk-write.py
   :start-after: start-bulk-write-unordered
   :end-before: end-bulk-write-unordered
   :language: python
   :copyable:

If any of the write operations in the preceding example fail, {+driver-short+}
reports the errors only after attempting all operations.

.. note::

   Unordered bulk operations do not guarantee order of execution. The order can

>    differ from the way you list them to optimize the runtime.

Reviewer: Earlier it says "arbitrary" order; does it actually do runtime optimization here?

Return Value
------------

The ``bulk_write()`` method returns a ``BulkWriteResult`` object. The
``BulkWriteResult`` object contains the following properties:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Property
     - Description

   * - ``acknowledged``
     - | Indicates if the server acknowledged the write operation.

   * - ``bulk_api_result``
     - | The raw bulk API result returned by the server.
         
   * - ``deleted_count``
     - | The number of documents deleted, if any.

   * - ``inserted_count``
     - | The number of documents inserted, if any.

   * - ``matched_count``
     - | The number of documents matched for an update, if applicable.

   * - ``modified_count``
     - | The number of documents modified, if any.

   * - ``upserted_count``
     - | The number of documents upserted, if any.

   * - ``upserted_ids``
     - | A map of operation index to the ``_id`` of the upserted documents, if

>          applicable.

Reviewer: I think there's a syntax error here but I don't know what it is. Maybe "operations indexed"?

Author: I took this from the API docs, but I'm pretty sure it's supposed to be a map that maps the index of the operation to the _id of the documents. I updated the wording a bit

Additional Information
----------------------

To learn how to perform individual write operations, see the following guides:

- :ref:`pymongo-write-insert`
- :ref:`pymongo-write-update`
- :ref:`pymongo-write-replace`

> - :ref:`pymongo-write-delete`

Author: These links will only work once this PR is merged

   :language: python
   :copyable:

To insert multiple documents, create an instance of ``InsertOne`` for each document.

Update Operations
~~~~~~~~~~~~~~~~~

To update a document, create an instance of ``UpdateOne`` and pass in
the following arguments:

> - A **query filter** that specifies the criteria used to match documents to in your collection

Reviewer: I think this technically works, but "to in" is a little weird

```suggestion
- A **query filter** that specifies the criteria used to match documents in your collection
```

> - The update operation you want to perform. For more information on update

Reviewer: ```suggestion
- The update operation you want to perform. For more information about update
```

   * - ``ordered``
     - | If ``True``, the driver performs the write operations in the order
         provided. If an error occurs, the remaining operations are not
         attempted. 
       |
       | If ``False``, the driver performs the operations in an

>          arbitrary order and attempts to perform each operation.

Reviewer: S: could change "each" to "all" to make slightly clearer

```suggestion
         arbitrary order and attempts to perform all operations.
```
